helix-core/src/main/java/org/apache/helix/controller/rebalancer/util/RebalanceScheduler.java

```diff
@@ -80,6 +80,7 @@ public void scheduleRebalance(HelixManager manager, String resource, long startT
     long delay = startTime - System.currentTimeMillis();
     if (delay < 0) {
       LOG.debug(String.format("Delay time is %s, will not be scheduled", delay));
+      return;
```

Comment:
Fixes #1325

Reply:
Good catch.

Reply:
I'm not convinced this should not be scheduled. I believe the expected behavior is triggering the rebalance immediately. In this case, we shall fix the log not the logic.

Overall, if the caller wants to schedule a rebalance, then it should be done. It is the caller responsibility to decide whether it is necessary. The scheduler should ensure it happens. Otherwise, we may miss the event by not doing anything. That will be a real bug.

Reply:
In my opinion, if delay is less than 0, we should not schedule rebalance. I will let Neal to convince you though. But anyways it is not the concern of this PR.

Reply:
The fix was done based on the expectation of this function, as seen in the comment: add a future rebalance task. In my opinion this function's duty is for future rebalance scheduling, therefore a start time in the past should be rejected as shown in the log. 

If a user wants to schedule an immediate rebalance, we have `scheduleOnDemandRebalance`. If a user is calling this function incorrectly (passing in a past time), then the problem should show up during tests (by not triggering rebalance) instead of hiding it and let it slip to production; if a user is using this function correctly and wants to schedule a rebalance in the future at time=x, but somehow this function is delayed and is processed at time=x+1, then should that rebalance still happen? I don't think we should assume it's okay, therefore it should be rejected. 

@jiajunwang Functionally I understand your point; code-quality-wise I think this function should do what it advertises. If there's a problem it should be caught by tests and the negative delay should be handled during development.

Reply:
1. If this change is not absolutely necessary for this PR, let's split.

2. Does it really cause any trouble for us? We should do the right thing, not what has been written down as a comment.
The potential problem I see here is that, as you mentioned, the possible gap between the timestamp generated and the time when this method starts processing. The delayed rebalancer has this potential issue if you return prematurely. So if you want to change it, then you need to "fix" the delayed rebalance. But back to the first point, I don't think it is a problem and the comment is the only thing we shall fix.

Anyway, please convince me with a scenario that without return, our logic will fail. Otherwise, #1325 does not look like an issue to me.

Reply:
The initial motivation was that 1 test case is failing. After digging around, it turns out that 1 negative delay was passed into the function, which removes a scheduled rebalance at a later time. A part of the fix was that the negative delay shouldn't be passed (it was unintentional), and the other part of the fix was #1325. 

I'm unable to produce any errors now that negative delays are not passed, so I can no longer provide such a scenario. 

#1325 should either be resolved by rejecting negative delay or fixing the description and expectation of this function. For now I'll remove it from this PR.